packages/flutter_tools/lib/src/android/android_studio.dart

```diff
@@ -0,0 +1,257 @@
+  }
+
+  factory AndroidStudio.fromHomeDot(Directory homeDotDir) {
+    String version = homeDotDir.basename.substring('.AndroidStudio'.length);
+    String installPath;
+    try {
+      installPath = fs
+          .file(fs.path.join(homeDotDir.path, 'system', '.home'))
+          .readAsStringSync();
+    } catch (e) {
+      // ignored
+    }
+    if (installPath != null && fs.isDirectorySync(installPath)) {
+      return new AndroidStudio(installPath, version: version);
+    }
+    return null;
+  }
+
+  String get gradlePath => _gradlePath;
+
+  String get gradleExecutable => fs.path
+      .join(_gradlePath, 'bin', platform.isWindows ? 'gradle.bat' : 'gradle');
+
+  bool get isValid => _isValid;
+
+  List<String> get validationMessages => _validationMessages;
+
+  @override
+  int compareTo(AndroidStudio other) {
+    int result = version.compareTo(other.version);
+    if (result == 0)
+      return directory.compareTo(other.directory);
+    return result;
+  }
+
+  /// Locates the newest, valid version of Android Studio.
+  static AndroidStudio latestValid() {
+    String configuredStudio = config.getValue('android-studio-dir');
+    if (configuredStudio != null) {
+      String configuredStudioPath = configuredStudio;
+      if (os.isMacOS && !configuredStudioPath.endsWith('Contents'))
+        configuredStudioPath = fs.path.join(configuredStudioPath, 'Contents');
+      return new AndroidStudio(configuredStudioPath,
+          configured: configuredStudio);
+    }
+
+    // Find all available Studio installations.
+    List<AndroidStudio> studios = allInstalled();
+    if (studios.isEmpty) {
+      return null;
+    }
+    studios.sort();
+    return studios.lastWhere((AndroidStudio s) => s.isValid,
+        orElse: () => null);
+  }
+
+  static List<AndroidStudio> allInstalled() =>
+      platform.isMacOS ? _allMacOS() : _allLinuxOrWindows();
+
+  static List<AndroidStudio> _allMacOS() {
+    List<FileSystemEntity> candidatePaths = <FileSystemEntity>[];
+
+    void _checkForStudio(String path) {
+      List<Directory> directories = fs
+          .directory(path)
+          .listSync()
+          .where((FileSystemEntity e) => e is Directory);
+      for (Directory directory in directories) {
+        if (directory.basename == 'Android Studio.app') {
+          candidatePaths.add(directory);
+        } else if (!directory.path.endsWith('.app')) {
+          _checkForStudio(directory.path);
+        }
+      }
+    }
+
+    _checkForStudio('/Applications');
+    _checkForStudio(fs.path.join(homeDirPath, 'Applications'));
+
+    String configuredStudioDir = config.getValue('android-studio-dir');
+    if (configuredStudioDir != null) {
+      FileSystemEntity configuredStudio = fs.file(configuredStudioDir);
+      if (configuredStudio.basename == 'Contents') {
+        configuredStudio = configuredStudio.parent;
+      }
+      if (!candidatePaths
+          .any((FileSystemEntity e) => e.path == configuredStudio.path)) {
```

Comment:
nit: your vertical alignment of the `.` is weird. I'd probably just put this all on one line, or wrap after the `any(` and before the matching `)` to make the structure clearer.